script.adzapper/service.py

```diff
@@ -0,0 +1,298 @@
+
+def getPlayer():
+    props = {'player': None, 'playerid': None, 'media': None, 'id': None, 'channel': None}
+    query = {
+            "method": "Player.GetActivePlayers",
+            }
+    res = jsonrpc(query)
+    if 'result' in res and res['result']:
+        res = res['result'][0]
+        props['player'] = res['type']
+        props['playerid'] = res['playerid']
+
+        query = {
+                "method": "Player.GetItem",
+                "params": {"properties": ["title", "season", "episode", "file", "channel"],
+                           "playerid": props['playerid']},
+                "id": "VideoGetItem"
+                }
+        res = jsonrpc(query)
+        if 'result' in res:
+            res = res['result'].get('item')
+            props['media'] = res['type']
+            props['channel'] = res['channel']            
+            if 'id' in res: props['id'] = res['id']
+    return props
+    
+def switchToChannelId(playerProperties, channelId, channel):
+
+    if playerProperties['player'] == 'audio' or (playerProperties['player'] == 'video' and playerProperties['media'] != 'channel'):
+
+        # stop all other players except pvr
+
+        log('player:%s media:%s @id:%s is running' % (playerProperties['player'], playerProperties['media'], playerProperties['playerid']))
+        query = {
+            "method": "Player.Stop",
+            "params": {"playerid": playerProperties['playerid']},
+        }
+        res = jsonrpc(query)
+        if 'result' in res and res['result'] == "OK":
+            handler.notifyLog('Player stopped')
+
+    log('Currently playing channelid %s (%s), switch to id %s' % (playerProperties['id'], playerProperties['channel'], channelId))
+    query = {
+        "method": "Player.Open",
+        "params": {"item": {"channelid": channelId}}
+    }
+    res = jsonrpc(query)
+    if 'result' in res and res['result'] == 'OK':
+        log('Switched to channel \'%s\'' % (channel))
+    else:
+        log('Couldn\'t switch to channel \'%s\'' % (channel))
+
+
+# Class to detect when something in the system has changed
+class ADZapperMonitor(xbmc.Monitor):
+    def onSettingsChanged(self):
+        log("Monitor: Notification of settings change received")
+        Settings.reloadSettings()
+        # Need to wait a couple of seconds for the changes to work
+        # through, reloading the keymap straight away didn't work
+        if not self.waitForAbort(2):
+            # Once the settings are reloaded, then we need to regenerate the
+            # keymap as that may have changes
+            keymap = KeyMapCtrl()
+            keymap.enableKeymap()
+            del keymap
+
+
+# Handles creating, updating and deleting the key map file
+class KeyMapCtrl():
+    KEY_MAP_SECTIONS = [
+        'global',
+        'VideoOSD',
+        'FullscreenLiveTV',
+    ]
+
+    def __init__(self):
+        self.keymapLocation = os.path.join(xbmc.translatePath('special://userdata/keymaps'), "adzapper_keymap.xml")
```

Comment:
Yep, I don't think automatically defining a keymap is allowed, it can create all sorts of abnormal behaviour. It is probably better to just import the keymap editor and guide the user to manually setting a keymap